### ISSUE TITLE
[Duende EF 6.1.0-preview.4] ConfigurationDbContext can't be created by DI container

### ISSUE THREAD

**Issue by marcindruzgala** [DuendeSoftware/Support#55]
**Which version of Duende IdentityServer are you using?**

`Duende.IdentityServer.EntityFramework 6.1.0-preview.4` - upgraded to preview in order to play around with server side session management

**Which version of .NET are you using?**
6.x.x

**Describe the bug**

This is similar to https://github.com/DuendeSoftware/IdentityServer/issues/723 but not exactly the same - issue with design time factory was solved with proposed solution in the issue.

Another issue that I think might be connected to changes done by @brockallen [here](https://github.com/DuendeSoftware/IdentityServer/pull/801/files).

We have following configuration db context setup:

```c#
public static IIdentityServerBuilder AddConfigurationStore(this IIdentityServerBuilder builder, string connectionString)
{
    builder.AddConfigurationStore<ConfigurationDbContext>(options =>
    {
        options.ResolveDbContextOptions = (sp, dbContextOptions) =>
        {
            dbContextOptions.UseNpgsql(connectionString, o =>
            {
                o.MigrationsAssembly(typeof(Program).Assembly.FullName);
            });
            dbContextOptions.UseApplicationServiceProvider(sp);
        };

        options.EnablePooling = true;
    });

    return builder;
}
```

Then in our program we're manually running migrations (not ideal, I know - that's not the point) like this:

```c#
public static async Task Main(string[] args)
{
    var host = CreateHostBuilder(args).Build();

    using var scope = host.Services.CreateScope();

    var configurationDbContext = scope.ServiceProvider.GetRequiredService<ConfigurationDbContext>();
    await configurationDbContext.Database.MigrateAsync();

    var persistedGrantDbContext = scope.ServiceProvider.GetRequiredService<PersistedGrantDbContext>();
    await persistedGrantDbContext.Database.MigrateAsync();

    ...
    await host.RunAsync();
}
```

But when I run our app we get an exception:

```c#
System.InvalidOperationException: The DbContext of type 'ConfigurationDbContext' cannot be pooled because it does not have a public constructor accepting a single parameter of type DbContextOptions or has more than one constructor.
   at Microsoft.EntityFrameworkCore.Internal.DbContextPool`1.CreateActivator(DbContextOptions`1 options)
   at Microsoft.EntityFrameworkCore.Internal.DbContextPool`1..ctor(DbContextOptions`1 options)
   ...
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
```

I have also tried changing `ResolveDbContextOptions` to `ConfigureDbContext` with basically the same code for it apart from
`UseApplicationServiceProvider` but the result is the same.

**Expected behavior**

ConfigurationDbContext is properly created.

**Comment by brockallen** [comment]
Hmm... do you know if the DI container you're using in this scenario is different than the one normally used in ASP.NET Core? The reason I ask is that I tested this in ASP.NET Core, and it's able to pick the right ctor based on what services are in DI, so it can choose the ctor with just the one param.

**Comment by marcindruzgala** [comment]
Are you also using db context pooling? Not sure if that makes any difference but the exception does say `The DbContext of type 'ConfigurationDbContext' cannot be pooled because...`

We're using the default ASP .NET Core DI container.

I'll try to reproduce it in some kind of sample app and link a repository here (hopefully this week :)).

**Comment by brockallen** [comment]
Hmm... I just ran our test EF host like this:

```
.AddConfigurationStore(options =>
            {
                options.ResolveDbContextOptions = (sp, dbContextOptions) =>
                {
                    dbContextOptions.UseSqlServer(connectionString, o =>
                    {
                        o.MigrationsAssembly(typeof(Program).Assembly.FullName);
                    });
                    dbContextOptions.UseApplicationServiceProvider(sp);
                };
                options.EnablePooling = true;
            })
```

And I can hit the discovery document without error.

**Comment by marcindruzgala** [comment]
I commented out the migrations code (so getting db context from DI container etc) in the Program class and it does work...I can authorize my clients and issue tokens etc

Could you try getting the db context in your program class, after host is built (but before you call `host.Run`):

```c#
var host = CreateHostBuilder(args).Build();
using var scope = host.Services.CreateScope();
var context = scope.ServiceProvider.GetRequiredService<ConfigurationDbContext>();
host.Run();
```

And see if that works for you?

**Comment by brockallen** [comment]
Ok, yes -- I can confirm what you're seeing. Not sure what I want to do about it yet... normally the migrations are best separate from the running web app, and thus you don't need pooling enabled in that one. 

Do you happen to know why runtime creation works, but the migrations fails? Seems really odd to me.

**Comment by brockallen** [comment]
> Hmm... thinking about this, it's likely due to DI lifetime issues (if they're doing pooling), meaning the instances are pooled, so they refuse to allow anything else other than the simple ctor to be used so no scoped dependencies are allowed.

But now I'm even more confused -- why does it work at runtime then?

**Comment by marcindruzgala** [comment]
> > it seems that somehow creating ConfigurationDbContext by calling GetRequiredService doesn't work but injecting it to another class works
> 
> Hmm... thinking about this, it's likely due to DI lifetime issues (if they're doing pooling), meaning the instances are pooled, so they refuse to allow anything else other than the simple ctor to be used so no scoped dependencies are allowed.

But we're doing exactly the same with 'our' db context, which is also pooled and it works :) It only happens with the `ConfigurationDbContext`.

Also pooled means that the underlying connection to the database is not being disposed but kept in the pool for later use, not the db context, right?

>But now I'm even more confused -- why does it work at runtime then?

I think I'm also very confused at this point :)

**Comment by brockallen** [comment]
> But we're doing exactly the same with 'our' db context

Does your DbContext take some extra param that comes from DI?

**Comment by brockallen** [comment]
Can you try adding this:

```
services.AddTransient<ConfigurationDbContext>(svcs => new ConfigurationDbContext(svcs.GetRequiredService<DbContextOptions<ConfigurationDbContext>>()));
```

**Comment by marcindruzgala** [comment]
That seems to fix it. I'm just wondering about registering that `dbContext` as transient? Shouldn't it be `Scoped`?

**Comment by brockallen** [comment]
Oh yea, that would be more appropriate. Good point.

**Comment by marcindruzgala** [comment]
Ok that kind of fixes that, I'm just wondering if registering it manually like that has any effect on the pooling of this specific db context?

Do you think it is something you'll be trying to fix or is it not really related to your implementation and I should look for help in the ef core repo?

**Comment by marcindruzgala** [comment]
The actual code throwing this exception comes from `DbContextPool`:

```c#
private static Func<DbContext> CreateActivator(DbContextOptions<TContext> options)
{
    var constructors = typeof(TContext).GetTypeInfo().DeclaredConstructors.Where(c => !c.IsStatic && c.IsPublic).ToArray();

    if (constructors.Length == 1)
    {
        var parameters = constructors[0].GetParameters();
        if (parameters.Length == 1
            && (parameters[0].ParameterType == typeof(DbContextOptions)
                || parameters[0].ParameterType == typeof(DbContextOptions<TContext>)))
        {
            return Expression.Lambda<Func<TContext>>(Expression.New(constructors[0], Expression.Constant(options))).Compile();
        }
    }

    throw new InvalidOperationException(CoreStrings.PoolingContextCtorError(typeof(TContext).ShortDisplayName()));
}
```

So it seems like, if pooling is enabled, db context can have only one public constructor that accepts only one argument of type `DbContextOptions`/`DbContextOptions<TContext>`.

I'm still very confused why it works when I just remove the migration code from `Program` class and then use any of the oidc endpoints. Then it's somehow able to create this context even with pooling enabled?

**Comment by brockallen** [comment]
Yes, not sure. It only throws on migrations but not actual runtime... but there are two ctors -- one with the single DbContextOptions and another with the extra param. I'm confused for sure.

**Comment by marcindruzgala** [comment]
Ok I think I understand why - correct me if I'm wrong.

In `IdentityServerEntityFrameworkBuilderExtensions` you have this extension method to register both `ConfigurationDbContext` and `PersistedGrantDbContext`. 

Now, if we take a closer look at the `AddOperationalDbContext<TContext>`:

```c#
// configuration of db context pool for PersistedGrantDbContext type
...

// then you register a scoped dependency for the interface `IPersistedGrantDbContext`
services.AddScoped<IPersistedGrantDbContext, TContext>();
```
From what I can see in some of the parts of your code (I haven't gone through all of it of course). You then inject `IPersistedGrantDbContext` into for example `PersistedGrantStore`. And that's why it works during runtime - because it DI is actually resolving _normal_ instance of this db context NOT the pooled one (which requires only 1 constructor) that is also being registered (but as a concrete implementation, not this interface).

Based on this I'm not really sure if configuring pooled context for your `DbContexts` actually makes any difference...your entire code is actually _not_ going to use db context pooling(?).

Could you verify if that is true? Or I'm I just talking gibberish right now?

**Comment by marcindruzgala** [comment]
For it to actually use db context pool you'd have to register it like MS guys are doing that in their `AddDbContextPool` code

MS code:
```c#
public static IServiceCollection AddDbContextPool<TContextService, TContextImplementation>(
    this IServiceCollection serviceCollection,
    Action<IServiceProvider, DbContextOptionsBuilder> optionsAction,
    int poolSize = DbContextPool<DbContext>.DefaultPoolSize)
    where TContextImplementation : DbContext, TContextService
    where TContextService : class
{
    Check.NotNull(serviceCollection, nameof(serviceCollection));
    Check.NotNull(optionsAction, nameof(optionsAction));

    AddPoolingOptions<TContextImplementation>(serviceCollection, optionsAction, poolSize);

    serviceCollection.TryAddSingleton<IDbContextPool<TContextImplementation>, DbContextPool<TContextImplementation>>();
    serviceCollection.TryAddScoped<IScopedDbContextLease<TContextImplementation>, ScopedDbContextLease<TContextImplementation>>();

    serviceCollection.TryAddScoped<TContextService>(
        sp => sp.GetRequiredService<IScopedDbContextLease<TContextImplementation>>().Context);

    if (typeof(TContextService) != typeof(TContextImplementation))
    {
        serviceCollection.TryAddScoped(p => (TContextImplementation)p.GetService<TContextService>()!);
    }

    return serviceCollection;
}
```

So as you can see if the `TContextService` is different from `TContextImplementation` (in your case `IPersistedGrantDbContext` vs `PersistedGrantDbContext`) they register it a bit differently.

So it would have to look more or less like this:

```c#
services.AddScoped<IPersistedGrantDbContext>(sp => sp.GetRequiredService<IScopedDbContextLease<PersistedGrantDbContext>>().Context);
```

**Comment by brockallen** [comment]
Hmm... yep, I think you have uncovered the mystery -- I did some local changed to confirm. That's interesting, so say the least. And so the conclusion here is that I think I need to revert the extra ctor change, which is a bummer for some folks. And I'll submit a PR to make these changes.

Many thanks!

**Comment by brockallen** [comment]
PR here: https://github.com/DuendeSoftware/IdentityServer/pull/893

Have a look, please.

**Comment by marcindruzgala** [comment]
Noice! Really appreciate the fast feedback and updates - looks like it should fix the issue 💯 

Just a question: why was the extra cctor needed?

**Comment by brockallen** [comment]
> Just a question: why was the extra cctor needed?

We had many reports of people having a hard time getting the options into DI, so they preferred the explicit ctor param approach.

**Comment by marcindruzgala** [comment]
@brockallen Just FYI - I was just going through on of the PRs and I don't know if it's going to be merged or not but I found this one also, that has the same issue: https://github.com/DuendeSoftware/IdentityServer/pull/860/files#diff-6aa80ca629263bdeeefe6ea23f72f66e5ae8ac9291e69479d017596ca4384580R87

**Comment by brockallen** [comment]
Yep, I got that one as well. And we released 6.1 today, FWIW.

Thx!
